RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc

```diff
@@ -0,0 +1,198 @@
+                                  const typename CAL::CaloRecHitSoATypeDevice::ConstView recHits,
+                                  reco::PFRecHitDeviceCollection::View pfRecHits,
+                                  uint32_t* __restrict__ denseId2pfRecHit,
+                                  uint32_t* __restrict__ num_pfRecHits) const {
+      // Strided loop over CaloRecHits
+      for (int32_t i : cms::alpakatools::elements_with_stride(acc, recHits.metadata().size())) {
+        // Check energy thresholds/quality cuts (specialised for HCAL/ECAL)
+        if (!applyCuts(recHits[i], params))
+          continue;
+
+        // Use atomic operation to determine index of the PFRecHit to be constructed
+        // The index needs to be unique and consequtive across all threads in all blocks.
+        // This is achieved using the alpaka::hierarchy::Blocks argument.
+        const uint32_t j = alpaka::atomicAdd(acc, num_pfRecHits, 1u, alpaka::hierarchy::Blocks{});
+
+        // Construct PFRecHit from CAL recHit (specialised for HCAL/ECAL)
+        constructPFRecHit(pfRecHits[j], recHits[i]);
+
+        // Fill denseId -> pfRecHit index map
+        denseId2pfRecHit[CAL::detId2denseId(pfRecHits.detId(j))] = j;
+      }
+    }
+
+    ALPAKA_FN_ACC static bool applyCuts(const typename CAL::CaloRecHitSoATypeDevice::ConstView::const_element rh,
+                                        const typename CAL::ParameterType::ConstView params);
+
+    ALPAKA_FN_ACC static void constructPFRecHit(
+        reco::PFRecHitDeviceCollection::View::element pfrh,
+        const typename CAL::CaloRecHitSoATypeDevice::ConstView::const_element rh);
+  };
+
+  template <>
+  ALPAKA_FN_ACC bool PFRecHitProducerKernelConstruct<HCAL>::applyCuts(
+      const typename HCAL::CaloRecHitSoATypeDevice::ConstView::const_element rh,
+      const HCAL::ParameterType::ConstView params) {
+    // Reject HCAL recHits below enery threshold
+    float threshold = 9999.;
+    const uint32_t detId = rh.detId();
+    const uint32_t depth = HCAL::getDepth(detId);
+    const uint32_t subdet = getSubdet(detId);
+    if (subdet == HcalBarrel) {
+      threshold = params.energyThresholds()[depth - 1];
+    } else if (subdet == HcalEndcap) {
+      threshold = params.energyThresholds()[depth - 1 + HCAL::kMaxDepthHB];
+    } else {
+      printf("Rechit with detId %u has invalid subdetector %u!\n", detId, subdet);
+    }
+    return rh.energy() >= threshold;
+  }
+
+  template <>
+  ALPAKA_FN_ACC bool PFRecHitProducerKernelConstruct<ECAL>::applyCuts(
+      const ECAL::CaloRecHitSoATypeDevice::ConstView::const_element rh, const ECAL::ParameterType::ConstView params) {
+    // Reject ECAL recHits below energy threshold
+    if (rh.energy() < params.energyThresholds()[ECAL::detId2denseId(rh.detId())])
+      return false;
+
+    // Reject ECAL recHits of bad quality
+    if ((ECAL::checkFlag(rh.flags(), ECAL::Flags::kOutOfTime) && rh.energy() > 2) ||
```

Comment:
why `rh.energy() > 2` ?

Reply:
This is taken from https://github.com/fllor/cmssw/blob/pfRecHitProducerAlpaka/RecoParticleFlow/PFClusterProducer/interface/PFRecHitQTests.h#L458